lib/srv/desktop/discovery.go

```diff
@@ -198,20 +194,53 @@ func (s *WindowsService) lookupDesktop(ctx context.Context, hostname string) ([]
 		return result
 	}
 
-	addrs, err := net.DefaultResolver.LookupNetIP(tctx, "ip4", hostname)
-	if err == nil && len(addrs) > 0 {
-		return stringAddrs(addrs), nil
-	}
-	if s.dnsResolver == nil {
-		return nil, trace.NewAggregate(err, trace.Errorf("DNS lookup for %q failed and there's no LDAP server to fallback to", hostname))
+	// kick off both DNS queries in parallel, but write results
+	// to separate channels so we can prefer the default resolver
+	defaultResult := make(chan []netip.Addr, 1)
+	ldapResult := make(chan []netip.Addr, 1)
+
+	go func() {
+		tctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+		defer cancel()
+		addrs, err := net.DefaultResolver.LookupNetIP(tctx, "ip4", hostname)
+		if err != nil {
+			s.cfg.Log.Debugf("DNS lookup for %q failed with default resolver: %v", hostname, err)
+		}
+		if len(addrs) > 0 {
+			defaultResult <- addrs
+		}
+	}()
+
+	if s.dnsResolver != nil {
+		go func() {
+			tctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+			defer cancel()
+			addrs, err := s.dnsResolver.LookupNetIP(tctx, "ip4", hostname)
+			if err != nil {
+				s.cfg.Log.Debugf("DNS lookup for %q failed with LDAP resolver: %v", hostname, err)
+			}
+			if len(addrs) > 0 {
+				ldapResult <- addrs
+			}
+		}()
```

Comment:
```suggestion
	const queryTimeout = time.Second * 5

	queryResolver := func(resolver *net.Resolver, resolverName string) chan []netip.Addr {
		resultChan := make(chan []netip.Addr, 1)

		if resolver != nil {
			go func() {
				tctx, cancel := context.WithTimeout(ctx, queryTimeout)
				defer cancel()
				addrs, err := resolver.LookupNetIP(tctx, "ip4", hostname)
				if err != nil {
					s.cfg.Log.Debugf("DNS lookup for %q failed with %s resolver: %v", resolverName, hostname, err)
				}
				if len(addrs) > 0 {
					resultChan <- addrs
				}
			}()
		}

		return resultChan
	}

	// kick off both DNS queries in parallel, but write results
	// to separate channels, so we can prefer the default resolver
	defaultResult := queryResolver(net.DefaultResolver, "default")
	ldapResult := queryResolver(s.dnsResolver, "LDAP")

	// wait 5 seconds for the default resolver to return
	select {
	case addrs := <-defaultResult:
		return stringAddrs(addrs), nil
	case <-s.cfg.Clock.After(queryTimeout):
	}
```
```suggestion
	// kick off both DNS queries in parallel, but write results
	// to separate channels so we can prefer the default resolver
	defaultResult := make(chan []netip.Addr, 1)
	ldapResult := make(chan []netip.Addr, 1)

	go func() {
		tctx, cancel := context.WithTimeout(ctx, 5*time.Second)
		defer cancel()
		addrs, err := net.DefaultResolver.LookupNetIP(tctx, "ip4", hostname)
		if err != nil {
			s.cfg.Log.Debugf("DNS lookup for %q failed with default resolver: %v", hostname, err)
		}
		if len(addrs) > 0 {
			defaultResult <- addrs
		}
	}()

	if s.dnsResolver != nil {
		go func() {
			tctx, cancel := context.WithTimeout(ctx, 5*time.Second)
			defer cancel()
			addrs, err := s.dnsResolver.LookupNetIP(tctx, "ip4", hostname)
			if err != nil {
				s.cfg.Log.Debugf("DNS lookup for %q failed with LDAP resolver: %v", hostname, err)
			}
			if len(addrs) > 0 {
				ldapResult <- addrs
			}
		}()
```

How do you feel about sharing the querying logic and introducing a constant for the timeout?